### DISCUSSION TITLE
Building SCIRun 5 beta Y on ubuntu 20.04

### DISCUSSION THREAD

**Discussion by peterrjohnston** [SCIInstitute/SCIRun#2196] [Q&A] [answered]
Hello,

I was wondering if I could get a bit help building SCIRun version 5, beta version Y, on ubuntu 20.04 LTS. There seems to be some problem with libboost and python as the process gets to the point: 

make[5]: *** No rule to make target '/home/peter/fromnet/SCIRun-5.0-beta.Y/bin/Externals/Source/Boost_external/lib/libboost_python-mt.a', needed by 'lib/libCore_Utils.so'.  Stop.

It is perhaps worth noting that I have the files libboost_python38.a and libboost_python38.so in the directory /usr/lib/x86_64-linux-gnu.

I have attached the complete output from build.sh as well as the CMakeCache.txt file.

I am running python version 3.8.5. Interestingly, the system builds perfectly OK if I set BUILD_WITH_PYTHON to OFF.

Any insights would be much appreciated. I can also provide you with more details of my system, but I am sure exactly what you need to know. Please don't hesitate to contact me if you require further information.

Thanks very much,

Peter.
[builderrors.txt](https://github.com/SCIInstitute/SCIRun/files/5539738/builderrors.txt)
[CMakeCache.txt](https://github.com/SCIInstitute/SCIRun/files/5539739/CMakeCache.txt)

**Comment by dcwhite** [comment]
Thanks! I've started working on this, since there are similar build errors related to boost/python on the latest Mac/XCode tools (XCode 12.2 and Catalina/Big Sur). The Github build didn't catch them since they're still running Ubuntu 18.04.

**Comment by dcwhite** [comment]
Update: I have a 20.04 build with python seemingly working on the Travis build virtual machine (see [here](https://travis-ci.org/github/SCIInstitute/SCIRun/builds/744532681)). It works with the clang compiler but not gcc. It worked with Python version 3.5.6 and 3.6.7 (which you select in the cmake setting `USER_PYTHON_VERSION`). I have not verified this locally however, but will attempt this soon. I did replicate the gcc build failure on my 20.04 machine, though, so you're not alone here. 

In the meantime, you could try building with the clang compiler--7.0.0 is what the Travis machine used.

**Comment by peterrjohnston** [comment]
Thanks Dan,

I am not exactly sure how to do what you are asking. At the moment, my machine says:

[peter:bin] > c++ --version
clang version 7.0.1-12 (tags/RELEASE_701/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

And:

[peter:SCIRun-5.0-beta.Y] > sudo update-alternatives --config c++
There are 2 choices for the alternative c++ (providing /usr/bin/c++).

  Selection    Path                Priority   Status
------------------------------------------------------------
* 0            /usr/bin/clang++-7   60        auto mode
  1            /usr/bin/clang++-7   60        manual mode
  2            /usr/bin/g++         20        manual mode

Using these options I get to exactly the same point in the build process as previously.

Any suggestions would be most welcome.

Thanks again,

Peter.

**Comment by dcwhite** [comment]
Can you paste the initial cmake output when switching compilers? It might be better to use separate build folders, so the cmake cache doesn't interfere. The cmake setting to change is CMAKE_CXX_COMPILER.

Clang should say 
```
-- The C compiler identification is Clang 7.0.0
-- The CXX compiler identification is Clang 7.0.0
-- Check for working C compiler: /usr/local/clang-7.0.0/bin/clang
-- Check for working C compiler: /usr/local/clang-7.0.0/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/clang-7.0.0/bin/clang++
-- Check for working CXX compiler: /usr/local/clang-7.0.0/bin/clang++ -- works
```

and gcc should say something like
```
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
```

Later today I am going to work on getting a clang build running locally, I will update this when I have it.

**Comment by peterrjohnston** [comment]
Thanks for continuing to look at this for me.

I get the following for the C compiler:

-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is Clang 7.0.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works

But I get nothing for gcc. 

Clearly, I have set up something incorrectly, but not sure what. Also, I am not sure how to change things as the output I get from c++ --version is given above.

Thanks,

Peter.

**Comment by dcwhite** [comment]
Update: I have a new Linux machine at home (with PopOs 20.10, a Ubuntu fork). I was able to get similar errors as you! So at least I can replicate this problem now more reliably. 

One thing I'm trying, but that didn't fix everything, is to force a clang build using some environment variables, like so:
```
$ export CC=/usr/bin/clang
$ export CXX=/usr/bin/clang++
$ cmake ..
-- The C compiler identification is Clang
-- The CXX compiler identification is Clang
```
I got this from this page https://stackoverflow.com/questions/7031126/switching-between-gcc-and-clang-llvm-using-cmake

**Comment by peterrjohnston** [comment]
Thanks for the update.  I tried your suggestion. Unfortunately, I didn't get any further than I did before.  It is still complaining about $SCIRun6_home/bin/Externals/Source/Boost_external/lib/libboost_python-mt.a being needed bu lib/libCore_Utils.so.

**Comment by dcwhite** [comment]
I think I have this build fixed on a branch, #2229. When that PR gets merged, I'll have you test again.

**Reply by dcwhite** [reply]
The Python-enabled build ought to work on Ubuntu 20.04 now. I tested with various versions of python, too; 3.8 is the default now.

**Comment by dcwhite** [comment]
v5.0-beta.Z is official, and should build on Ubuntu 20.04 among other Linux versions. Please give it a try

**Comment by peterrjohnston** [comment, marked as answer]
Thanks Dan,
It seems to work perfectly now. Thanks very much for your efforts.
